### ISSUE TITLE
Does SimpleExec swallow non-zero exit errors?

### ISSUE THREAD

**Issue by javabrett** [github/git-lfs#1183]
In the context of #1177, I was testing what happens if the new `git lfs install` option `--system` is run by a non-root user.  Expect failure of course, since the resulting `git config --system` command won't have access to `/etc/gitconfig`, and will return non-zero.

I was working on the `Set*` commands, but couldn't get them to pass-through an error.  Then I looked at [`SimpleExec`](https://github.com/github/git-lfs/blob/master/subprocess/subprocess.go#L13)

```
func SimpleExec(name string, args ...string) (string, error) {
        tracerx.Printf("run_command: '%s' %s", name, strings.Join(args, " "))
        cmd := ExecCommand(name, args...)

        output, err := cmd.Output()
        if _, ok := err.(*exec.ExitError); ok {
                return "", nil
        }
        if err != nil {
                return fmt.Sprintf("Error running %s %s", name, args), err
        }

        return strings.Trim(string(output), " \n"), nil
}
```

There's a decent chance I am reading this incorrectly, but it looks like this exec-wrapper intentionally discards any non-catastrophic process-error.  Unless the process failed to launch or abended, other failures e.g. a non-zero exit code will result in an `ExitError`, which is being converted to no-output-no-error `return "", nil`.

Is that analysis right?  And if so, is there a reason for that ... is it that so many `git` commands return non-zero e.g. when tests run, and we don't care to handle the error?

I changed this logic and of course a bunch of tests break, so I thought I'd check if there is any history behind this.

In short, how do we detect that `git config --system ...` as non-root, returned exit code `255` and string `error: could not lock config file /etc/gitconfig: Permission denied`?

**Comment by technoweenie** [comment]
It does, and it probably shouldn't. In the short term, I'd recommend using `subprocess.ExecCommand()` directly. Long term, we should probably deprecate `SimpleExec`, and create a similar helper that returns non zero exit codes with stderr as the error message.

**Comment by javabrett** [comment]
I'm going to propose a tweak to `SimpleExec` to accept a boolean `swallowErr` and react accordingly.  Moving to `ExecCommand` is a good suggestion and I looked into it, but it turns out that the wrapper is useful in formatting the forked-process's stdout/err etc. into an error when required.

**Comment by javabrett** [comment]
Dropped the optional/extra boolean parameter approach, in-favour of just making sure `""` is still returned as `Output` whenever there is a `ExitError`.  So the change is now to not nil-ify the error, and improve the contents of the error to show additional detail from the process stderr.  Requires code pull-down from Go 1.6 API change adding ExitError.Stderr, which can be reverted once >= Go 1.6.

Required a error-handling fix picked-up by `ls-file` empty repo test, which exposed a previously never-true error check.

**Comment by chasescot** [comment]
now that's what I call a helpful information thank you


Sent from my Samsung Galaxy smartphone.-------- Original message --------From: Brett Randall <notifications@github.com> Date: 2016-04-28  5:49 PM  (GMT-07:00) To: github/git-lfs <git-lfs@noreply.github.com> Subject: Re: [github/git-lfs] Does SimpleExec swallow non-zero exit errors?
  (#1183) 
Dropped the optional/extra boolean parameter approach, in-favour of just making sure "" is still returned as Output whenever there is a ExitError.  So the change is now to not nil-ify the error, and improve the contents of the error to show additional detail from the process stderr.  Requires code pull-down from Go 1.6 API change adding ExitError.Stderr, which can be reverted once >= Go 1.6.

Required a error-handling fix picked-up by ls-file empty repo test, which exposed a previously never-true error check.
